templates/decays.cpp.in

```diff
@@ -0,0 +1,301 @@
+
+using namespace @ModelName@_cxx_diagrams;
+using namespace @ModelName@_cxx_diagrams::fields;
+namespace info = @ModelName@_info;
+using namespace std::complex_literals;
+
+const @ModelName@_decay_table& CLASSNAME::get_decay_table() const
+{
+   return decay_table;
+}
+
+void CLASSNAME::set_loop_corrections(const Decays_loop_corrections& loop_corrections_)
+{
+   loop_corrections = loop_corrections_;
+}
+
+const Decays_loop_corrections& CLASSNAME::get_loop_corrections() const
+{
+   return loop_corrections;
+}
+
+const Decays_problems& CLASSNAME::get_problems() const
+{
+   return problems;
+}
+
+void CLASSNAME::clear_problems()
+{
+   problems.clear();
+}
+
+void CLASSNAME::clear()
+{
+   clear_problems();
+   decay_table.clear();
+}
+
+void CLASSNAME::calculate_decays()
+{
+#ifdef ENABLE_THREADS
+@callAllDecaysFunctionsInThreads@
+#else
+@callAllDecaysFunctions@
+#endif
+}
+
+double get_alphas(context_base const&  context)
+{
+   return Sqr(context.model.get_@gs_name@())/(4.*Pi);
+}
+
+double get_alpha(context_base const&  context)
+{
+   return Sqr(unit_charge(context))/(4.*Pi);
+}
+
+/* 1-loop BSM amplitudes
+ *
+ * Notes:
+ *    - Since FS removes a factor of i from every vertex we reintroduce it
+ *      calls to one loop amplitudes
+ */
+
+@calcAmplitudeSpecializationDefs@
+
+// -------- specializations for decays needing higher order SM corrections --------
+
+@create_SM_particle_usings@
+
+namespace {
+
+unsigned int number_of_active_flavours(double m)
+{
+   /*
+   if (m < qedqcd.displayMbMb(){
+      return 4.0;
+   } else if (m < qedqcd.displayPoleMt()) {
+      return 5.0;
+   } else {
+      return 6.0;
+   }
+   */
+   return 5;
+}
+
+// eq. 2.7 of hep-ph/0503173
+double calc_A(double b)
+{
+   return (1 + b * b) *
+             (4. * dilog((1 - b) / (1 + b)) + 2. * dilog(-(1 - b) / (1 + b)) -
+              3. * std::log((1 + b) / (1 - b)) * std::log(2 / (1 + b)) -
+              2. * std::log((1 + b) / (1 - b)) * std::log(b)) -
+          3. * b * std::log(4. / (1 - b * b)) - 4. * b * std::log(b);
+}
+
+// eq. 2.6 of hep-ph/0503173
+double calc_DeltaH(double b)
+{
+   return calc_A(b)/b + 1./(16*b*b*b) * (
+      3 + 34*b*b - 13*Power4(b)) * log((1+b)/(1-b)) + 3./(8.*b*b)*(7*b*b-1);
+}
+
+// eq. 2.6 of hep-ph/0503173
+double calc_DeltaAH(double b)
+{
+   return calc_A(b)/b + 1./(16*b) * (19 + 2*b*b + 3*Power4(b))
+      * log((1+b)/(1-b)) + 3./8.*(7-b*b);
+}
+
+// 2 and higher order
+double calc_deltaqq(double alpha_s_red, double Nf)
+{
+   return
+      // order alphas are taken into account with mass sependence somewhere alse
+      0. * 17./3. * alpha_s_red
+      + (35.94 - 1.36 * Nf) * Sqr(alpha_s_red)
+      + (164.14 - 25.77 * Nf + 0.259 * Sqr(Nf)) * Power3(alpha_s_red)
+      + (39.34 - 220.9 * Nf + 9.685 * Sqr(Nf) - 0.0205 * Power3(Nf)) * Power4(alpha_s_red);
+}
+
+// eq. 2.31 of hep-ph/0503172
+double RT(double x) {
+   BOOST_ASSERT_MSG(x<=1, "This function");
```

Comment:
Please don't use an assertion here, as it may lead to resource leaks.
I'd propose to either
* throw an exception (domain error or similar) or
* print an error message using the macros from `logger.hpp` and return NaN.